src/hypothesis/strategies.py

```diff
@@ -836,31 +836,75 @@ def dm_func(denom):
 
 @cacheable
 @defines_strategy
-def decimals(
-    min_value=None, max_value=None, allow_nan=None, allow_infinity=None
-):
-    """Generates instances of decimals.Decimal.
+def decimals(min_value=None, max_value=None,
+             allow_nan=True, allow_infinity=None, places=None):
+    """Generates instances of decimals.Decimal, which may be:
 
-    - If min_value is not None, all values will be >= min_value.
-    - If max_value is not None, all values will be <= max_value.
-    - If min_value or max_value is not None, it is an error to enable
-      allow_nan.
-    - If both min_value and max_value are not None, it is an error to enable
-      allow_infinity.
+    - A finite rational number, between ``min_value`` and ``max_value``.
+    - Not a Number, if ``allow_nan`` is truthy.
+    - Positive or negative infinity, if ``max_value`` and ``min_value``
+      respectively are None, and ``allow_infinity`` is not False.  (None means
+      "allow infinity, unless excluded by the min and max values")
 
-    Where not explicitly ruled out by the bounds, all of infinity, -infinity
-    and NaN are possible values generated by this strategy.
+    Note that where floats have one `NaN` value, Decimals have four: signed,
+    and either *quiet* or *signalling*.  See `the decimal module docs
+    <https://docs.python.org/3/library/decimal.html#special-values>`_ for
+    more information on special values.
 
     """
-    floats_strategy = floats(
-        min_value=min_value, max_value=max_value,
-        allow_nan=allow_nan, allow_infinity=allow_infinity)
-    fractions_strategy = fractions(min_value=min_value, max_value=max_value)
-
-    return (
-        floats_strategy.map(float_to_decimal) |
-        fractions_strategy.map(lambda f: Decimal(f.numerator) / f.denominator)
-    )
+    # Convert min_value and max_value to Decimal values, and validate args
+    check_valid_integer(places)
+    if places is not None and places < 0:
+        raise InvalidArgument('cannot have negative number of places (%r)')
+    try:
+        min_value = Decimal(min_value)
+    except:
+        pass
+    try:
+        max_value = Decimal(max_value)
+    except:
+        pass
+    if isinstance(min_value, Decimal) and not min_value.is_finite():
+        raise InvalidArgument('min_value=%r is not a finite bound' % min_value)
+    if isinstance(max_value, Decimal) and not max_value.is_finite():
+        raise InvalidArgument('max_value=%r is not a finite bound' % max_value)
+    check_valid_bound(min_value, 'min_value')
+    check_valid_bound(max_value, 'max_value')
+    check_valid_interval(min_value, max_value, 'min_value', 'max_value')
+    if allow_infinity is None:
+        allow_infinity = min_value is not None or max_value is not None
+    elif allow_infinity and min_value is not None and max_value is not None:
+        raise InvalidArgument('Cannot allow infinity between finite bounds')
+    # Set up a strategy for finite decimals
+    if places is not None:
```

Comment:
This is not a request for changes, but I'm curious if you tried just quantizing after the value was generated and if so what problems that caused.